test/OpenIddict.Server.Tests/OpenIddictServerExtensionsTests.TestCases.cs

```diff
@@ -0,0 +1,417 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OpenIddict.Server.Tests
+{
+    public partial class OpenIddictServerExtensionsTests
+    {
+        private static IEnumerable<AddServer_RegistersDefaultHandler_TestCase> GetOpenIdDictServerHandlerRegistrationTestCases()
```

Comment:
I get the intent, but I suspect maintaining these test cases will be painful in the long term.

Instead, couldn't we just make sure all the `OpenIddictServerHandlerDescriptor`s contained in `OpenIddictServerHandlers.DefaultHandlers` array have their `ServiceDescriptor` registered by `AddServer()`? It would make all this stuff unnecessary for your `AddServer_RegistersDefaultHandler` test.

Reply:
So, my initial concern was that if you had a test failure come up, the test name would be unhelpful in tracking down what registration was missing - this was particularly true for types which end up being a generic.

One possible way these tests could be rewritten:
```csharp
public static IEnumerable<object[]> GetAddServer_RegistersDefaultHandler_V2TestCases()
{
	foreach (var descriptor in OpenIddictServerHandlers.DefaultHandlers)
	{
		yield return new object[] { descriptor };
	}
}

[Theory]
[MemberData(nameof(GetAddServer_RegistersDefaultHandler_V2TestCases))]
public void AddServer_RegistersDefaultHandler_v2(OpenIddictServerHandlerDescriptor descriptor)
{
	//Arrange
	var services = new ServiceCollection();
	var builder = new OpenIddictBuilder(services);

	//Act
	builder.AddServer();

	//Assert
	Assert.Contains(services, service => service.Lifetime == descriptor.ServiceDescriptor.Lifetime &&
			service.ServiceType == descriptor.ServiceDescriptor.ServiceType &&
			service.ImplementationType == descriptor.ServiceDescriptor.ImplementationType);
}
```
 Here's an example output shown for a failure (I commented out the line in AddServer that registers all the default handlers):
> AddServer_RegistersDefaultHandler_v2(descriptor: OpenIddictServerHandlerDescriptor { ContextType = typeof(OpenIddict.Server.OpenIddictServerEvents+ApplyAuthorizationResponseContext), FilterTypes = [], Order = -2147381648, ServiceDescriptor = ServiceType: OpenIddict.Server.OpenIddictServerHandlers+Authentication+AttachResponseState Lifetime: Singleton ImplementationType: OpenIddict.Server.OpenIddictServerHandlers+Authentication+AttachResponseState, Type = BuiltIn }) [0:00.000] Failed: Assert.Contains() Failure
> Assert.Contains() Failure
> Not found: (filter expression)
> In value:  ServiceCollection [ServiceType: Microsoft.Extensions.Options.IOptions`1[TOptions] Lifetime: Singleton ImplementationType: Microsoft.Extensions.Options.OptionsManager`1[TOptions], ServiceType: Microsoft.Extensions.Options.IOptionsSnapshot`1[TOptions] Lifetime: Scoped ImplementationType: Microsoft.Extensions.Options.OptionsManager`1[TOptions], ServiceType: Microsoft.Extensions.Options.IOptionsMonitor`1[TOptions] Lifetime: Singleton ImplementationType: Microsoft.Extensions.Options.OptionsMonitor`1[TOptions], ServiceType: Microsoft.Extensions.Options.IOptionsFactory`1[TOptions] Lifetime: Transient ImplementationType: Microsoft.Extensions.Options.OptionsFactory`1[TOptions], ServiceType: Microsoft.Extensions.Options.IOptionsMonitorCache`1[TOptions] Lifetime: Singleton ImplementationType: Microsoft.Extensions.Options.OptionsCache`1[TOptions], ...]
>    at OpenIddict.Server.Tests.OpenIddictServerExtensionsTests.AddServer_RegistersDefaultHandler_v2(OpenIddictServerHandlerDescriptor descriptor) in ...\test\OpenIddict.Server.Tests\OpenIddictServerExtensionsTests.cs:line 215

Another downside of this approach, is that if a change were to happen where one of the handlers in question was removed from the DefaultHandlers array, no test would fail, as that test case would be removed.

Reply:
> One possible way these tests could be rewritten:

I like this approach 👍 

> Another downside of this approach, is that if a change were to happen where one of the handlers in question was removed from the DefaultHandlers array, no test would fail, as that test case would be removed.

I'd expect this case to be caught by the integration tests, where a missing or invalid registration should result in an invalid pipeline.